### PR TITLE
add a hook in Session to support the proof generation progress bar

### PR DESCRIPTION
**Pull request by dimitrysuen**
It would be great for users to view the proof generation progress. We put a hook in `Session` to make it extensible.

Also, I tested it locally. Here is the screenshot:
screenshot

### CODE REVIEW

File: risc0/zkvm/src/session.rs [hunk 1]
```diff
@@ -92,13 +97,24 @@ pub struct Segment {
...
 }
 
+/// The Events of [Session]
+pub trait Events {
+    /// Fired before the proving of a segment starts.
+    fn on_pre_prove_segment(&self, _: &Segment) {}
+
+    /// Fired after the proving of a segment ends.
+    fn on_post_prove_segment(&self, _: &Segment) {}
+}
```
Comment (flaub) [inline risc0/zkvm/src/session.rs:107]:
```suggestion
pub trait SessionEvents {
    /// Fired before the proving of a segment starts.
    fn on_pre_prove_segment(&self, segment: &Segment) {}

    /// Fired after the proving of a segment ends.
    fn on_post_prove_segment(&self, segment: &Segment) {}
}
```

**Comment by flaub** [comment]
This is pretty nice, thank you!

Could you add a test for this so that we don't accidentally break this feature in the future?

File: risc0/zkvm/src/session.rs (hunk 1 above)
Comment (dimitrysuen) [reply risc0/zkvm/src/session.rs]: Thanks for the suggestion. We've renamed `Events` and `_`.

**Comment by dimitrysuen** [comment]
> This is pretty nice, thank you!
> 
> Could you add a test for this so that we don't accidentally break this feature in the future?

Yeah. We added the test for the hook.

File: risc0/zkvm/src/session.rs [hunk 2]
```diff
@@ -92,13 +97,25 @@ pub struct Segment {
     pub insn_cycles: usize,
 }
 
+/// The Events of [Session]
+pub trait SessionEvents {
+    /// Fired before the proving of a segment starts.
+    // The awkward let _ is required to suppress the unused variables warning
+    fn on_pre_prove_segment(&self, segment: &Segment) { let _ = segment; }
```
Comment (flaub) [inline risc0/zkvm/src/session.rs:104]:
```suggestion
    #[allow(unused)]
    fn on_pre_prove_segment(&self, segment: &Segment) { let _ = segment; }
```

Comment (dimitrysuen) [reply risc0/zkvm/src/session.rs]: Fixed. Thanks.

**Pull request closed without merging**